data/transactions/application.go

```diff
@@ -0,0 +1,776 @@
+	// key in the key/value store, there can be at most one delta per key.
+	// Therefore the order that the deltas are applied does not matter.
+	for key, valueDelta := range stateDelta {
+		switch valueDelta.Action {
+		case basics.SetUintAction:
+			kv[key] = basics.TealValue{
+				Type: basics.TealUintType,
+				Uint: valueDelta.Uint,
+			}
+		case basics.SetBytesAction:
+			kv[key] = basics.TealValue{
+				Type:  basics.TealBytesType,
+				Bytes: valueDelta.Bytes,
+			}
+		case basics.DeleteAction:
+			delete(kv, key)
+		default:
+			return fmt.Errorf("unknown delta action %d", valueDelta.Action)
+		}
+	}
+	return nil
+}
+
+// applyStateDeltas applies a basics.EvalDelta to the app's global key/value
+// store as well as a set of local key/value stores. If this function returns
+// an error, the transaction must not be committed.
+func (ac *ApplicationCallTxnFields) applyStateDeltas(evalDelta basics.EvalDelta, params basics.AppParams, creator, sender basics.Address, balances Balances, appIdx basics.AppIndex, errIfNotApplied bool) error {
+	/*
+	 * 1. Apply GlobalState delta (if any), allocating the key/value store
+	 *    if required.
+	 */
+
+	proto := balances.ConsensusParams()
+	if len(evalDelta.GlobalDelta) > 0 {
+		// Clone the parameters so that they are safe to modify
+		params = params.Clone()
+
+		// Allocate GlobalState if necessary. We need to do this now
+		// since an empty map will be read as nil from disk
+		if params.GlobalState == nil {
+			params.GlobalState = make(basics.TealKeyValue)
+		}
+
+		// Check that the global state delta isn't breaking any rules regarding
+		// key/value lengths
+		err := evalDelta.GlobalDelta.Valid(&proto)
+		if err != nil {
+			if !errIfNotApplied {
+				return nil
+			}
+			return fmt.Errorf("cannot apply GlobalState delta: %v", err)
+		}
+
+		// Apply the GlobalDelta in place on the cloned copy
+		err = applyDelta(params.GlobalState, evalDelta.GlobalDelta)
+		if err != nil {
+			return err
+		}
+
+		// Make sure we haven't violated the GlobalStateSchema
+		err = params.GlobalState.SatisfiesSchema(params.GlobalStateSchema)
+		if err != nil {
+			if !errIfNotApplied {
+				return nil
+			}
+			return fmt.Errorf("GlobalState for app %d would use too much space: %v", appIdx, err)
+		}
+	}
+
+	/*
+	 * 2. Apply each LocalState delta, fail fast if any affected account
+	 *    has not opted in to appIdx or would violate the LocalStateSchema.
+	 *    Don't write anything back to the cow yet.
+	 */
+
+	changes := make(map[basics.Address]basics.AppLocalState, len(evalDelta.LocalDeltas))
+	for accountIdx, delta := range evalDelta.LocalDeltas {
+		// LocalDeltas are keyed by account index [sender, tx.Accounts[0], ...]
+		addr, err := ac.AddressByIndex(accountIdx, sender)
+		if err != nil {
+			return err
+		}
+
+		// Ensure this is not a duplicate address in case we were
+		// passed an invalid EvalDelta
+		_, ok := changes[addr]
+		if ok {
```

Comment:
Is it okay to have duplicates of zero-deltas?

Reply:
I went ahead and just disallowed zero-length deltas. There is no reason to support them since they are never generated.